### PR TITLE
DISPATCH-1586 - Added allocator metrics

### PR DESCRIPTION
**Pull request by fgiorgetti**
Needs to be reviewed and metric must have the correct suffixes, following
prometheus conventions (https://prometheus.io/docs/practices/naming/).

Please review and let me know if the approach below seems valid or not.

### CODE REVIEW

File: src/http-libwebsockets.c
Comment (jiridanek) [inline src/http-libwebsockets.c:620]:
```suggestion
        *position += lws_snprintf((char*) *position, end - *position, "%s:%s_bytes %" PRIu64 "\n", name, field.name, field.value);
```

Comment (jiridanek) [reply src/http-libwebsockets.c:620]: Not all platforms define uint64_t as unsigned long int, https://issues.apache.org/jira/browse/DISPATCH-1669

File: src/http-libwebsockets.c
Comment (jiridanek) [inline src/http-libwebsockets.c:653]: magic number is magic :D

File: src/http-libwebsockets.c
Comment (fgiorgetti) [reply src/http-libwebsockets.c:620]: Ok!

**Comment by jiridanek** [comment]
Can anyone please comment on the "Please review and let me know if the approach below seems valid or not." part? CC @grs

File: src/http-libwebsockets.c
Comment (jiridanek) [reply src/http-libwebsockets.c:653]: please explain magic number, consider making it an int constant

**Comment by grs** [comment]
Looks ok to me

File: src/http-libwebsockets.c
Comment (fgiorgetti) [reply src/http-libwebsockets.c]: @jiridanek finally updated as suggested. Please feel free to review.

File: src/http-libwebsockets.c
Comment (jiridanek) [inline src/http-libwebsockets.c:704]: I'm wondering about the log levels, `QD_LOG_DEBUG` for success seems high to me (I'd go with TRACE) and `QD_LOG_DEBUG` for the failure here could be too low (I'd give it INFO). But I did not examine qdrouterd logging comprehensively, so maybe I am just no properly attuned.

File: src/http-libwebsockets.c
Comment (jiridanek) [inline src/http-libwebsockets.c:661]: looks a bit fragile, but ok

**Review by jiridanek — APPROVED** [review]
This is tested by the existing test for metrics? Since it is only adding more metrics...

Do the new metrics follow Prometheus guidelines, as you mention in first comment, or that is still something to check?

lgtm, but this is something I have little experience with...

**Comment by jiridanek** [comment]
One more thought. From the README

```
-DQD_MEMORY_STATS=
------------------

Dispatch will track memory pool usage statistics if this is enabled.```

So the statistics may not be always available, depending on the way Dispatch is compiled. Can that break your code, or is that then always going to report 0?

File: src/http-libwebsockets.c
Comment (fgiorgetti) [reply src/http-libwebsockets.c:704]: Makes sense to me as well.

**Comment by fgiorgetti** [comment]
> One more thought. From the README
> 
> ```
> -DQD_MEMORY_STATS=
> ------------------
> 
> Dispatch will track memory pool usage statistics if this is enabled.
> ```
> 
> So the statistics may not be always available, depending on the way Dispatch is compiled. Can that break your code, or is that then always going to report 0?

Allocator metrics get reported as 0. I see that `qdmanage` works just fine, but `qdstat` is throwing an error in such case:

```
$ qdstat -m
2021-02-04 15:08:19.288818 UTC
Standalone_v++jCPCjYiQcGo3

TypeError: unsupported operand type(s) for *: 'int32' and 'NoneType'
```

**Comment by ganeshmurthy** [comment]
Please do not merge this PR until after we start the vote for 1.15

**Comment by jiridanek** [comment]
I'll merge and I'll create a Jira for broken qdstat

**Comment by grs** [comment]
Personally I think having a broken qdstat on master is something to avoid. We also need to make sure that we still have compatability with older qdstat (and vice versa).

In my opinion it is better to get a solution to any consequent problems like this *before* merging.

**Comment by jiridanek** [comment]
> In my opinion it is better to get a solution to any consequent problems like this _before_ merging.

My understanding was that qdstat was already broken on master (when compiled with that nondefault option) and the exception is not caused by this PR. Cc @fgiorgetti can you clarify?

**Comment by jiridanek** [comment]
@grs I checked, the bug DISPATCH-1958 was already present, before this PR got merged. I tested at commit e78a18ae.

**Comment by grs** [comment]
@jiridanek sorry, I misunderstood. If the matser is already broken no issue with the merge.

**Pull request merged**
